### PR TITLE
Editorial: Refactor SetNumberFormatDigitOptions

### PR DESCRIPTION
**Pull request by sffc**
This is an editorial change intended to re-structure the SetNumberFormatDigitOptions abstract operation to reduce diffs in the NumberFormat v3 proposal.

@ryzokuken, you've worked extensively on this abstract operation before, so I would like you to review it to make sure that it is, indeed, editorial.

### CODE REVIEW

File: spec/numberformat.html [hunk 1]
```diff
@@ -22,27 +22,37 @@ <h1>SetNumberFormatDigitOptions ( _intlObj_, _options_, _mnfdDefault_, _mxfdDefa
...
+        1. Let _hasFd_ be _mnfd_ is not *undefined* or _mxfd_ is not *undefined*.
+        1. Let _needSd_ be _hasSd_.
+        1. Let _needFd_ be not _hasSd_ and _notation_ is not *"compact"*.
+        1. If _needSd_, then
+          1. If _hasSd_, then
+            1. Let _mnsd_ be ? DefaultNumberOption(_mnsd_, 1, 21, 1).
+            1. Let _mxsd_ be ? DefaultNumberOption(_mxsd_, _mnsd_, 21, 21).
+            1. Set _intlObj_.[[MinimumSignificantDigits]] to _mnsd_.
+            1. Set _intlObj_.[[MaximumSignificantDigits]] to _mxsd_.
+          1. Else,
```
[the remark this answers is not in the document]
Comment (ryzokuken) [reply spec/numberformat.html:35]: Maybe it is there just to align this with the next condition? Does that achieve much though?

[the remark this answers is not in the document]
Comment (sffc) [reply spec/numberformat.html:35]: The difference becomes apparent in NFv3.

[the remark this answers is not in the document]
Comment (ryzokuken) [reply spec/numberformat.html:35]: @sffc ah, okay, I suspected so... perhaps it is okay to drop these additional conditions here and re-add them in NFv3?

**Comment by anba** [comment]
> 1. Let _hasSd_ be _mnsd_ is not *undefined* or _mxsd_ is not *undefined*.

ECMASpeak doesn't have implicit to boolean conversions in this context. So this needs to be more like:

> 1. If _mnsd_ is not *undefined* or _mxsd_ is not *undefined*, set _hasSd_ to *true*.
> 1. Else, set _hasSd_ to *false*.

File: spec/numberformat.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (sffc) [reply spec/numberformat.html:35]: I have a slight preference to keep the extra variables because (1) they more clearly signify intent and (2) they reduce the size of the NFv3 diff, but I'm willing to change it if you feel more strongly.

[the remark this answers is not in the document]
Comment (ryzokuken) [reply spec/numberformat.html:35]: No, I don't feel that strongly.

**Comment by ryzokuken** [comment]
@sffc could you please address the review?

File: spec/numberformat.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (sffc) [reply spec/numberformat.html]: I resolved this a bit by denesting the _hasSd_ construction.

**Review by sffc — COMMENTED** [review]
How's this?

File: spec/numberformat.html [hunk 2]
```diff
@@ -22,27 +22,35 @@ <h1>SetNumberFormatDigitOptions ( _intlObj_, _options_, _mnfdDefault_, _mxfdDefa
         1. Let _mnsd_ be ? Get(_options_, *"minimumSignificantDigits"*).
         1. Let _mxsd_ be ? Get(_options_, *"maximumSignificantDigits"*).
         1. Set _intlObj_.[[MinimumIntegerDigits]] to _mnid_.
-        1. If _mnsd_ is not *undefined* or _mxsd_ is not *undefined*, then
-          1. Set _intlObj_.[[RoundingType]] to ~significantDigits~.
+        1. If _mnsd_ is not *undefined* or _mxsd_ is not *undefined*, set _hasSd_ to *true*; else, set _hasSd_ to *false*.
+        1. If _mnfd_ is not *undefined* or _mxfd_ is not *undefined*, set _hasFd_ to *true*; else, set _hasFd_ to *false*.
```
Comment (ryzokuken) [inline spec/numberformat.html:26]: Any reason why both of these conditions are on a single line instead of being split across two?

File: spec/numberformat.html [hunk 3]
```diff
@@ -22,27 +22,35 @@ <h1>SetNumberFormatDigitOptions ( _intlObj_, _options_, _mnfdDefault_, _mxfdDefa
...
+        1. Let _needSd_ be _hasSd_.
+        1. If not _hasSd_ and _notation_ is not *"compact"*, set _needFd_ to *true*; else, set _needFd_ to *false*.
```
Comment (ryzokuken) [inline spec/numberformat.html:28]: Same as above.

**Review by ryzokuken — APPROVED** [review]
Just two editorial questions, everything else LGTM.

File: spec/numberformat.html (hunk 2 above)
Comment (sffc) [reply spec/numberformat.html:26]: Because I find it more readable to write

1. If _condition_, set _a_ to *true*; else, set _a_ to *false*.
1. If _condition_, set _b_ to *true*; else, set _b_ to *false*.
1. If _condition_, set _c_ to *true*; else, set _c_ to *false*.
1. If _condition_, set _d_ to *true*; else, set _d_ to *false*.

than to write

1. If _condition_, set _a_ to *true*.
1. Else, set _a_ to *false*.
1. If _condition_, set _b_ to *true*.
1. Else, set _b_ to *false*.
1. If _condition_, set _c_ to *true*.
1. Else, set _c_ to *false*.
1. If _condition_, set _d_ to *true*.
1. Else, set _d_ to *false*.

But if you disagree, or if my syntax is not normal for the spec, I will change it.

Comment (ryzokuken) [reply spec/numberformat.html:26]: I agree that it's more readable, but I am unsure if (1) this is a spec convention (2) it even matters.

Maybe we can ask the 262 editors about it?

Comment (sffc) [reply spec/numberformat.html:26]: @syg @anba thoughts?

Comment (sffc) [reply spec/numberformat.html:26]: Ping @anba @syg

Comment (anba) [reply spec/numberformat.html:26]: Both alternatives are valid and are used in ECMA-262 and ECMA-402. 

I think as a rule of thumb, I normally just try to follow the usage in the surrounding code resp. operations. Except when the condition is overly long or in a deeply nested step, so that the short form would either be rendered across multiple lines or when readability suffers.

**Comment by gibson042** [comment]
Since you're making changes to SetNumberFormatDigitOptions, can you also improve its description and eliminate the initial assertion steps by adopting a structured header (cf. [#594](https://github.com/tc39/ecma402/pull/594/files#diff-f5e04a56730a1b4629cb21e028b69b5773841156890909110bc52b03c8e944a8))?
```
    <emu-clause id="sec-setnfdigitoptions" type="abstract operation">
      <h1>
        SetNumberFormatDigitOptions (
          _intlObj_: an Object,
          _options_: an Object,
          _mnfdDefault_: a Number,
          _mxfdDefault_: a Number,
          _notation_: a String,
        )
      </h1>
      <dl class="header">
        <dt>description</dt>
        <dd>It [description of what it does].</dd>
      </dl>
      <emu-alg>…</emu-alg>
```

**Review by gibson042 — CHANGES_REQUESTED** [review]
This change appears to be normative: https://jsfiddle.net/x0jq9drc/

Example observable impact:
SetNumberFormatDigitOptions(_intlObj_, {maximumFractionDigits: 0}, 0, 3, "compact")
```diff
 [[MinimumIntegerDigits]]: 1
-[[RoundingType]]: ~fractionDigits~
-[[MinimumFractionDigits]]: 0
-[[MaximumFractionDigits]]: 0
+[[RoundingType]]: ~compactRounding~
```

File: spec/numberformat.html [hunk 4]
```diff
@@ -23,26 +23,43 @@ <h1>SetNumberFormatDigitOptions ( _intlObj_, _options_, _mnfdDefault_, _mxfdDefa
...
         1. If _mnsd_ is not *undefined* or _mxsd_ is not *undefined*, then
-          1. Set _intlObj_.[[RoundingType]] to ~significantDigits~.
+          1. Set _hasSd_ to *true*.
+        1. Else,
+          1. Set _hasSd_ to *false*.
+        1. If _mnfd_ is not *undefined* or _mxfd_ is not *undefined*, then
+          1. Set _hasFd_ to *true*.
+        1. Else,
+          1. Set _hasFd_ to *false*.
+        1. Let _needSd_ be _hasSd_.
```
[the remark this answers is not in the document]
Comment (sffc) [reply spec/numberformat.html:33]: I like the parallelism between [has/need]Fd with [has/need]Sd.

Also, in NumberFormat v3, _needSd_ no longer equals _hasSd_.  I could merge the variables for now, but doing so will increase the diff relative to NFv3.

File: spec/numberformat.html [hunk 5]
```diff
@@ -23,26 +23,43 @@ <h1>SetNumberFormatDigitOptions ( _intlObj_, _options_, _mnfdDefault_, _mxfdDefa
         1. Let _mxsd_ be ? Get(_options_, *"maximumSignificantDigits"*).
         1. Set _intlObj_.[[MinimumIntegerDigits]] to _mnid_.
         1. If _mnsd_ is not *undefined* or _mxsd_ is not *undefined*, then
-          1. Set _intlObj_.[[RoundingType]] to ~significantDigits~.
-          1. Let _mnsd_ be ? DefaultNumberOption(_mnsd_, 1, 21, 1).
-          1. Let _mxsd_ be ? DefaultNumberOption(_mxsd_, _mnsd_, 21, 21).
+          1. Let _hasSd_ to *true*.
```
Comment (ptomato) [inline spec/numberformat.html:26]: (Drive-by comment as this popped up in my notifications:)
```suggestion
          1. Let _hasSd_ be *true*.
```

Comment (sffc) [reply spec/numberformat.html]: Thanks for catching this :)

**Comment by sffc** [comment]
> Since you're making changes to SetNumberFormatDigitOptions, can you also improve its description and eliminate the initial assertion steps by adopting a structured header (cf. [#594](https://github.com/tc39/ecma402/pull/594/files#diff-f5e04a56730a1b4629cb21e028b69b5773841156890909110bc52b03c8e944a8))?

Done!

> This change appears to be normative: https://jsfiddle.net/x0jq9drc/
> 
> Example observable impact:
> SetNumberFormatDigitOptions(_intlObj_, {maximumFractionDigits: 0}, 0, 3, "compact")
> 
> ```diff
>  [[MinimumIntegerDigits]]: 1
> -[[RoundingType]]: ~fractionDigits~
> -[[MinimumFractionDigits]]: 0
> -[[MaximumFractionDigits]]: 0
> +[[RoundingType]]: ~compactRounding~
> ```

Your JSFiddle is outstanding!  I made one diff in https://github.com/tc39/ecma402/pull/575/commits/dd7d0cf2ae2dea21e8cfbe48869592e1fb08fcf4 and managed to make your test case pass.  Thank you so much!  (Is it possible to get something like this into Test262? @ryzokuken)

File: spec/numberformat.html [hunk 6]
```diff
@@ -23,26 +30,43 @@ <h1>SetNumberFormatDigitOptions ( _intlObj_, _options_, _mnfdDefault_, _mxfdDefa
...
+          1. Let _hasSd_ be *false*.
+        1. If _mnfd_ is not *undefined* or _mxfd_ is not *undefined*, then
+          1. Let _hasFd_ be *true*.
+        1. Else,
+          1. Let _hasFd_ be *false*.
+        1. Let _needSd_ be _hasSd_.
+        1. If _hasSd_ is *false* and (_hasFd_ is *true* or _notation_ is not *"compact"*), then
+          1. Let _needFd_ be *true*.
+        1. Else,
+          1. Let _needFd_ be *false*.
```
Comment (gibson042) [inline spec/numberformat.html:44]: I don't think we express things with parentheses like this in spec text.
```suggestion
        1. If _hasSd_ is *true*, or _hasFd_ is *false* and _notation_ is *"compact"*, then
          1. Let _needFd_ be *false*.
        1. Else,
          1. Let _needFd_ be *true*.
```

File: spec/numberformat.html [hunk 7]
```diff
@@ -23,26 +30,43 @@ <h1>SetNumberFormatDigitOptions ( _intlObj_, _options_, _mnfdDefault_, _mxfdDefa
...
+            1. If _mnfd_ is *undefined*, set _mnfd_ to min(_mnfdDefault_, _mxfd_).
+            1. Else if _mxfd_ is *undefined*, set _mxfd_ to max(_mxfdDefault_, _mnfd_).
+            1. Else if _mnfd_ is greater than _mxfd_, throw a *RangeError* exception.
+            1. Set _intlObj_.[[MinimumFractionDigits]] to _mnfd_.
+            1. Set _intlObj_.[[MaximumFractionDigits]] to _mxfd_.
+          1. Else,
+            1. Set _intlObj_.[[MinimumFractionDigits]] to _mnfdDefault_.
+            1. Set _intlObj_.[[MaximumFractionDigits]] to _mxfdDefault_.
+        1. If _needSd_ or _needFd_, then
+          1. If _hasSd_, then
```
Comment (gibson042) [inline spec/numberformat.html:64]:
```suggestion
        1. If _needSd_ is *true* or _needFd_ is *true*, then
          1. If _hasSd_ is *true*, then
```
But for that matter, this part seems hard to follow. Broader suggestion:
```
        1. If _needSd_ is *false* and _needFd_ is *false*, then
            1. Set _intlObj_.[[RoundingType]] to ~compactRounding~.
        1. Else if _hasSd_ is *true*, then
            1. Set _intlObj_.[[RoundingType]] to ~significantDigits~.
        1. Else,
            1. Set _intlObj_.[[RoundingType]] to ~fractionDigits~.
```

**Comment by sffc** [comment]
Ping

**Review by ryzokuken — APPROVED** [review]

**Comment by ryzokuken** [comment]
@gibson042 could you please give this another review? It would be great to merge this in ASAP in order to move things along for NF v3.
